docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
-.. currentmodule:: torch.sparse
+.. currentmodule:: torch
 
 .. _sparse-docs:
 
 torch.sparse
 ============
 
-.. warning::
+Torch supports sparse tensors which provide memory-efficient storage
+and processing algorithms of tensors for which the majority of
+elements have the same value (fill value). This allows saving storage
+and processing resources when the size and sparsity levels of tensors
+are high.  Otherwise, for small-sized or low-sparsity tensors using
+strided tensors is likely the most efficient approach.
 
-    This API is in beta and may change in the near future.
+.. _sparse-fill-value-docs:
 
-Torch supports sparse tensors in COO(rdinate) format, which can
-efficiently store and process tensors for which the majority of elements
-are zeros.
+Fill value
+++++++++++
 
-A sparse tensor is represented as a pair of dense tensors: a tensor
-of values and a 2D tensor of indices.  A sparse tensor can be constructed
-by providing these two tensors, as well as the size of the sparse tensor
-(which cannot be inferred from these tensors!)  Suppose we want to define
-a sparse tensor with the entry 3 at location (0, 2), entry 4 at
-location (1, 0), and entry 5 at location (1, 2).  We would then write:
+In Torch, the fill value is not specified explicitly but is implicitly
+interpreted by sparse algorithms. The implicit interpretation may vary
+in-between algorithms of different fields and applications. The most
+commonly assumed fill value is zero and it is assumed for all tools
+that are available via :mod:`torch` namespace. When some sparse tensor
+functionality assumes a non-zero or indefinite fill value, the
+corresponding tools are provided via :mod:`torch.sparse`
+namespace. Please see the documentation strings of the corresponding
+functions for a particular definition of the fill value.
 
-    >>> i = torch.LongTensor([[0, 1, 1],
-                              [2, 0, 2]])
-    >>> v = torch.FloatTensor([3, 4, 5])
-    >>> torch.sparse.FloatTensor(i, v, torch.Size([2,3])).to_dense()
-     0  0  3
-     4  0  5
-    [torch.FloatTensor of size 2x3]
+.. _sparse-hybrid-docs:
 
-Note that the input to LongTensor is NOT a list of index tuples.  If you want
+Hybrid tensors
```

Comment:
How should the reader thing of hybrid tensors vs. sparse formats? Can the sparse portion be ANY sparse format? Can the "dense" or "strided" portion be any strided format?

Reply:
It seems like this section should, like fill_value, also follow the description of formats.

Reply:
Hybrid tensors are now described as an extension of a sparse tensor with scalar elements to a sparse tensor with tensor elements.
Yes, in principle, the sparse portion can be any sparse format. The generalization of a sparse tensor with XYZ format to hybrid XYZ tensor is straightforward: one must allow the values to be multi-dimensional tensors rather than scalar tensors.